pkg/controller/master/upgrade/upgrade_controller.go

```diff
@@ -831,3 +836,59 @@ func (h *upgradeHandler) startVM(ctx context.Context, vm *kubevirtv1.VirtualMach
 		Do(ctx)
 	return res.Error()
 }
+
+// checkLogReadyCondition times out LogReady condition, thus moving on to the next step if upgradeLogCondLogReadyTimeout
+// time has passed since UpgradeLog's logging infrastructure setup started
+func (h *upgradeHandler) checkLogReadyCondition(upgrade *harvesterv1.Upgrade) (*harvesterv1.Upgrade, error) {
+	upgradeConfig, err := settings.DecodeConfig[settings.UpgradeConfig](settings.UpgradeConfigSet.Get())
+	if err != nil {
+		logrus.Errorf("Failed to get UpgradeConfig")
+		return upgrade, err
+	}
+
+	timeoutStr := upgradeConfig.LogReadyTimeout
+	timeout, err := strconv.Atoi(timeoutStr)
+	if err != nil {
+		return upgrade, fmt.Errorf("invalid value for image preload timeout: %s", timeoutStr)
+	}
+	timeoutDuration := time.Duration(timeout) * time.Minute
+
+	if timeoutStr == "" || timeoutDuration < 1*time.Minute || timeoutDuration > 20*time.Minute {
+		logrus.Warnf("invalid logReadyTimeout must be between 1 to 20 minutes, given: %s", timeoutStr)
+		logrus.Infof("Using default timeoutStr: %v", settings.UpgradeLogReadyTimeoutDefault)
+		timeoutDuration = settings.UpgradeLogReadyTimeoutDefault
+	}
+
+	ts := harvesterv1.LogReady.GetLastUpdated(upgrade)
+	t, err := time.Parse(time.RFC3339, ts)
+	if err != nil {
+		logrus.Errorf("Failed to parse last updated time: %v", err)
+		return upgrade, nil
+	}
+	if time.Since(t) >= timeoutDuration {
+		toUpdate := upgrade.DeepCopy()
+		upgradeLog, err := h.upgradeLogCache.Get(util.HarvesterSystemNamespaceName, upgrade.Status.UpgradeLog)
+		if err != nil {
+			logrus.Errorf("Failed to get upgrade log: %v", err)
+			setLogReadyCondition(toUpdate, corev1.ConditionFalse, "UpgradeLogGetFailed", "Failed to get upgrade log")
+			return h.upgradeClient.Update(toUpdate)
+		}
+		logrus.Infof("Timed out waiting for UpgradeLog %q's %q condition to pass", upgradeLog.Name, harvesterv1.LogReady)
+
+		toUpdateUpgradeLog := upgradeLog.DeepCopy()
+		harvesterv1.InfraReady.SetStatus(toUpdateUpgradeLog, string(metav1.ConditionFalse))
+		harvesterv1.InfraReady.Reason(toUpdateUpgradeLog, "Timeout")
+		harvesterv1.InfraReady.Message(toUpdateUpgradeLog, "Timed out creating logging infrastructure")
+		_, err = h.upgradeLogClient.Update(toUpdateUpgradeLog)
+		if err != nil {
+			logrus.Errorf("Failed to update upgradeLog: %v", err)
+			return upgrade, nil
+		}
```

Comment:
Similar question here. Should we do `setUpgradeCompletedCondition` here as well?

Reply:
Thanks for explaining. Makes sense. Done.